sql: drop index in chunks

vivekmenezes opened the pull request:
This change drops an index in chunks using the new kv chunk API
that returns a resume-span after every chunk is deleted. The
resume-span is used on the next iteration of a chunked deletion.

Added a test for the chunked deletion of an interleaved index.
#7499

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8870)

<!-- Reviewable:end -->

vivekmenezes committed:
sql: drop index in chunks

This change drops an index in chunks using the new kv chunk API
that returns a resume-span after every chunk is deleted. The
resume-span is used on the next iteration of a chunked deletion.

Added a test for the chunked deletion of an interleaved index.

paperstreet commented:
Great stuff

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/backfill.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLnxhxXoFZHRbHui1x:-KQLnxhxXoFZHRbHui1y:b9tzndr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/backfill.go#L398)):*
> ```Go
> 				// Short circuit the truncation if the table has been deleted.
> 				if tableDesc.Deleted() {
> 					// Mark done.
> ```

This comment is not helpful, please remove.

---

*[sql/drop_test.go, line 239 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLqTjAgNF66KHW0Mj3:-KQLqTjAgNF66KHW0Mj4:bp2v73z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/drop_test.go#L239)):*
> ```Go
> 	// Bulk insert.
> 	maxValue := 2*sql.IndexBackfillChunkSize + 1
> 	insert := fmt.Sprintf(`INSERT INTO t.kv VALUES (%d, %d)`, 0, maxValue)
> ```

uses a `bytes.Buffer` for constructing these. Here and in the other test

---

*[sql/drop_test.go, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLqfzSTMav-Sl74qY5:-KQLqfzTl4E-sQsus5jE:b-ihcc59) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/drop_test.go#L240)):*
> ```Go
> 	maxValue := 2*sql.IndexBackfillChunkSize + 1
> 	insert := fmt.Sprintf(`INSERT INTO t.kv VALUES (%d, %d)`, 0, maxValue)
> 	for i := 1; i <= maxValue; i++ {
> ```

Why not 0 to maxValue-1?

---

*[sql/tablewriter.go, line 652 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLpiMCAQ7thOOVP43A:-KQLpiMCAQ7thOOVP43B:b-ecn6y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/tablewriter.go#L652)):*
> ```Go
> 
> // deleteIndex runs the kv operations necessary to delete all kv entries in the
> // given index. This may require a scan.
> ```

I might be worth extending this doc a bit to explain `resume` and the return `Span`

---

*[sql/tablewriter.go, line 654 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLolOzLO1w-QJWXw19:-KQLolP-e580nhhHEgTv:bs3igyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/tablewriter.go#L654)):*
> ```Go
> // given index. This may require a scan.
> func (td *tableDeleter) deleteIndex(
> 	ctx context.Context, idx *sqlbase.IndexDescriptor, resume roachpb.Span, bound int64,
> ```

I was unclear to me what`bound`did until I followed it down in the code, so I'm not a fan of this variable name. I think we usually call this `limit`

---

*[sql/tablewriter.go, line 685 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLpWPEBp8-0kIzh74F:-KQLpWPEBp8-0kIzh74G:b-gmthso) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/tablewriter.go#L685)):*
> ```Go
> 	}
> 	if l := len(td.b.Results); l != 1 {
> 		panic(fmt.Sprintf("%d results returned", l))
> ```

`, expected 1`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-:-KQLrXgck-PQ0Zoyv9bd:b-9g0l9g)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Thanks for taking a look. It's ready for another review.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[sql/backfill.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLnxhxXoFZHRbHui1x:-KQMGqcq2qqf1EfD4bwN:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/backfill.go#L398)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> This comment is not helpful, please remove.
</details>
removed

---

*[sql/drop_test.go, line 239 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLqTjAgNF66KHW0Mj3:-KQMEnPPWu2g7l6EXMAM:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/drop_test.go#L239)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> uses a `bytes.Buffer` for constructing these. Here and in the other test
</details>
done

---

*[sql/drop_test.go, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLqfzSTMav-Sl74qY5:-KQMEodrLP6bGO7F64De:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/drop_test.go#L240)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Why not 0 to maxValue-1?
</details>
fixed

---

*[sql/tablewriter.go, line 652 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLpiMCAQ7thOOVP43A:-KQMF80XsyYY2HvUeqzW:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/tablewriter.go#L652)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> I might be worth extending this doc a bit to explain `resume` and the return `Span`
</details>
done

---

*[sql/tablewriter.go, line 654 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLolOzLO1w-QJWXw19:-KQMF94Kdca1YuaEkBBD:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/tablewriter.go#L654)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> I was unclear to me what`bound`did until I followed it down in the code, so I'm not a fan of this variable name. I think we usually call this `limit`
</details>
done

---

*[sql/tablewriter.go, line 685 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLpWPEBp8-0kIzh74F:-KQMFi50jLeEbsIrp9UN:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/tablewriter.go#L685)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> `, expected 1`
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-:-KQMH9EKOBhlAPBB9-Bl:b8unw1r)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[sql/drop_test.go, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLqfzSTMav-Sl74qY5:-KQMIjnHlBtks5FNFTfx:b-ltm162) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/drop_test.go#L240)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> fixed
</details>
Oops, I was actually just curious if there was some specific reason. Either is fine.

If you're going to do 0 to maxValue, then `for i := 0; i < maxValue; i++` and `fmt.Sprintf(... maxValue...)` is idiomatic

---

*[sql/tablewriter.go, line 654 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQMJGNsym2AFvcGsYIS:-KQMJGNthFaueZL9Ac5D:b-2dwh3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec1ffab27388a93ab83bf54a3cd6cc205e6f9c3b/sql/tablewriter.go#L654)):*
> ```Go
> // given index. This may require a scan.
> //
> // resume = is the resume-span which should be used for the index deletion
> ```

I've never seen this `=` in docs before. It reads just fine without it

---

*[sql/tablewriter.go, line 659 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQMJX7z3qkCMQtp6OXX:-KQMJX7z3qkCMQtp6OXY:budx542) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec1ffab27388a93ab83bf54a3cd6cc205e6f9c3b/sql/tablewriter.go#L659)):*
> ```Go
> // span is returned.
> //
> // limit = is a limit on either the number of keys or table-rows deleted in
> ```

which is it a limit on? we should standardize on one of them, probably keyval entries

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-:-KQMJg17zIK6f4jnr2T5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[sql/drop_test.go, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQLqfzSTMav-Sl74qY5:-KQMKqRwJzcK3Hyxf4Cw:bimpe6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b7c5905ee97d14eaf75cc910b3c59a4bdc53556/sql/drop_test.go#L240)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Oops, I was actually just curious if there was some specific reason. Either is fine.
> 
> If you're going to do 0 to maxValue, then `for i := 0; i < maxValue; i++` and `fmt.Sprintf(... maxValue...)` is idiomatic
</details>
It was copied from another part of the codebase, but on review I felt that using 0 to maxValue -1 made sense here.

---

*[sql/tablewriter.go, line 654 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQMJGNsym2AFvcGsYIS:-KQML6Vtywz3Gtf1uz7H:brpmqbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec1ffab27388a93ab83bf54a3cd6cc205e6f9c3b/sql/tablewriter.go#L654)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> I've never seen this `=` in docs before. It reads just fine without it
</details>
removed 

---

*[sql/tablewriter.go, line 659 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870#-KQMJX7z3qkCMQtp6OXX:-KQMLVXPhcmT_RsY04_r:b1ou8b5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec1ffab27388a93ab83bf54a3cd6cc205e6f9c3b/sql/tablewriter.go#L659)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> which is it a limit on? we should standardize on one of them, probably keyval entries
</details>
Actually here we can make it a limit on the number of index entries that get deleted. Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8870)*
<!-- Sent from Reviewable.io -->
